pkg/lib/infra/redis/hub.go

```diff
@@ -0,0 +1,228 @@
+package redis
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/go-redis/redis/v8"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/log"
+)
+
+type SubscriberID struct {
+	Counter int64
+}
+
+func (i *SubscriberID) Next() int64 {
+	for {
+		curr := atomic.LoadInt64(&i.Counter)
+		if atomic.CompareAndSwapInt64(&i.Counter, curr, curr+1) {
+			return curr
+		}
+	}
+}
+
+type Subscriber struct {
+	ChannelName  string
+	SubscriberID int64
+	// The drop pattern
+	// See https://www.ardanlabs.com/blog/2017/10/the-behavior-of-channels.html
+	MessageChannel chan *redis.Message
+}
+
+// ControlMessageJoin is essentially Subscriber.
+type ControlMessageJoin = Subscriber
+
+// ControlMessageLeave is basically Subscriber without MessageChannel.
+type ControlMessageLeave struct {
+	ChannelName  string
+	SubscriberID int64
+}
+
+// PubSub is an actor wrapping redis.PubSub
+type PubSub struct {
+	PubSub     *redis.PubSub
+	Subscriber map[string][]Subscriber
+	// Mailbox handles 3 messages.
+	// 1. When the mailbox is closed, that means the actor should die.
+	// 2. ControlMessageJoin
+	// 3. ControlMessageLeave
+	Mailbox chan interface{}
+}
+
+// Hub aims to multiplex multiple subscription to Redis PubSub channel over a single connection.
+type Hub struct {
+	Pool         *Pool
+	SubscriberID *SubscriberID
+	Logger       *log.Logger
+
+	mutex  sync.RWMutex
+	pubsub map[string]*PubSub
+}
+
+func NewHub(pool *Pool, lf *log.Factory) *Hub {
+	h := &Hub{
+		Pool:         pool,
+		SubscriberID: &SubscriberID{},
+		Logger:       lf.New("redis-hub"),
+		pubsub:       make(map[string]*PubSub),
+	}
+	return h
+}
+
+func (h *Hub) Subscribe(
+	cfg *config.RedisConfig,
+	credentials *config.RedisCredentials,
+	channelName string,
+) (chan *redis.Message, func()) {
+	pubsub := h.getPubSub(cfg, credentials)
+
+	id := h.SubscriberID.Next()
+	// The drop pattern
+	// See https://www.ardanlabs.com/blog/2017/10/the-behavior-of-channels.html
+	c := make(chan *redis.Message, 10)
+
+	pubsub.Mailbox <- ControlMessageJoin{
```

Comment:
A bit of nitpick: this may have same 'send to closed channel' problem when subscription & pubsub actor mailbox close is performed concurrently. I suppose this case would not happen as it is now, since we never close the pubsub actor actively (is it? what about abnormal closure of redis conn?).

Conceptually, we have three types of actors here:
- `Hub`: one per Authgear server instance, supervise a set of `PubSub`,
- `PubSub`: one per Redis connection, supervise a set of `Subscriber`,
- `Subscriber`: one per WebSocket connection.

To be more robust, we may send the `Subscriber` actor reference through `Hub`, which dispatch it to the correct `PubSub` instance. The death of `PubSub` actors would be performed by the supervisor (`Hub`) after being notified by the child `PubSub` actor when connection failed.

Reply:
I converted Hub to an actor.